docs/v1-proposed-api.md

```diff
@@ -0,0 +1,145 @@
+# Proposed Changes for Hive V1 API
+
+## ClusterDeployment
+
+ 1. Require an InstallConfig ConfigMap, referenced in a new ClusterDeployment.Spec field, to pass directly to installer.
+   * GOAL: Allow quicker exposure to new Installer features without needing to mirror in Hive's API.
+   * GOAL: Remaining ClusterDeployment.Spec should be things that can be modified post-install.
+   * Should Hive parse the InstallConfig to obtain information it needs, eliminating the need for the caller to duplicate information in ClusterDeployment and InstallConfig?
+     * Or should Hive be explicitly given all the information it needs to adopt and maintain a cluster in it's ClusterDeployment, requiring the caller to duplicate a fair portion of InstallConfig in each.
+     * Should we use an InstallConfig if given one, otherwise generate?
+   * Add Spec.InstallConfigMap reference.
+   * Likely cannot embed pull secret and ssh key into the InstallConfig in a ConfigMap.
+     * Should we require the InstallConfig in a Secret instead?
+     * Should we template in the secret values on the fly?
+   * It does not appear that we can remove much of our API. Only Spec.Networking can go. We still need the platform for region, the pull secret + ssh key, the machine sets for on-going management and syncing. There is actually not that much config exposed otherwise.
+     * Doing this would still expose installer features faster, however the caller would have to be aware of two APIs instead of just one, including multiple versions of InstallConfig, and duplicate a bit of information in both.
+ 1. Remove deprecated Status.Installed in favor of newer Spec.Installed.
+ 1. Drop ClusterDeployment.Spec.Images, this is unused and unnecessary.
+ 1. Rename ClusterDeployment.Spec.SSHKey to ClusterDeployment.Spec.SSHKeySecret to better match all other secret references.
+ 1. Allow adoption of a cluster without requiring creation of a ClusterProvision:
+   * If we didn't provision it we should not need a ClusterProvision.
+   * Move refs to secrets with admin kubeconfig/password into Spec instead of Status.
+   * Move Status.ClusterID/Infra ID to Spec instead of Status.
+ 1. Add support for bare metal provisioning.
+   * This is modelled as another platform in InstallConfig. No reason we shouldn't do the same.
+
+## ClusterProvision
+
+ 1. Drop PodSpec, generate on the fly? (note: this is not within the Spec, external, possibly already generated on fly)
+
+## ClusterDeprovisionRequest
+
+ 1. Rename ClusterDeprovision, to better match ClusterProvision.
+
+## SyncSet + SelectorSyncSet
+
+ 1. Drop support for ApplyOnce patch mode.
+
+## HiveConfig
+
+ 1. Adjust HiveConfig API to map managed DNS domains to cloud provider they should be used for.
```

Comment:
Or, hear me out here, remove support for managed DNS. It is not something that fits well as a feature of Hive. It is something that is needed by one user of Hive and could easily reside outside of Hive.

Reply:
For posterity we discussed, this could be made to be an external component but probably not feasible at this time, and still something we consider under the purview of Hive, automating all aspects needed for provisioning OpenShift at scale.